datarepo-clienttests/src/main/java/utils/KubernetesClientUtils.java

```diff
@@ -0,0 +1,264 @@
+package utils;
+
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.GoogleCredentials;
+import io.kubernetes.client.openapi.ApiClient;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.Configuration;
+import io.kubernetes.client.openapi.apis.AppsV1Api;
+import io.kubernetes.client.openapi.apis.CoreV1Api;
+import io.kubernetes.client.openapi.models.V1Deployment;
+import io.kubernetes.client.openapi.models.V1DeploymentList;
+import io.kubernetes.client.openapi.models.V1DeploymentSpec;
+import io.kubernetes.client.openapi.models.V1Pod;
+import io.kubernetes.client.openapi.models.V1PodList;
+import io.kubernetes.client.util.ClientBuilder;
+import io.kubernetes.client.util.KubeConfig;
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import runner.config.ServerSpecification;
+
+// TODO: add try/catch for refresh token around all utils methods
+public final class KubernetesClientUtils {
+
+  private static int maximumSecondsToWaitForReplicaSetSizeChange = 500;
+  private static int secondsIntervalToPollReplicaSetSizeChange = 5;
+
+  public static final String componentLabel = "app.kubernetes.io/component";
+  public static final String apiComponentLabel = "api";
+
+  private static CoreV1Api kubernetesClientCoreObject;
+  private static AppsV1Api kubernetesClientAppsObject;
+
+  private KubernetesClientUtils() {}
+
+  public static CoreV1Api getKubernetesClientCoreObject() {
+    return kubernetesClientCoreObject;
+  }
+
+  public static AppsV1Api getKubernetesClientAppsObject() {
+    return kubernetesClientAppsObject;
+  }
+
+  /**
+   * Build the singleton Kubernetes client objects. This method should be called once at the
+   * beginning of a test run, and then all subsequent fetches should use the getter methods instead.
+   *
+   * @param server the server specification that points to the relevant Kubernetes cluster
+   */
+  public static void buildKubernetesClientObject(ServerSpecification server) throws Exception {
+    // call the fetchGKECredentials script that uses gcloud to generate the kubeconfig file
+    List<String> scriptArgs = new ArrayList<>();
+    scriptArgs.add("tools/fetchGKECredentials.sh");
+    scriptArgs.add(server.clusterShortName);
+    scriptArgs.add(server.region);
+    scriptArgs.add(server.project);
+    ProcessUtils.executeCommand("sh", scriptArgs);
+
+    // path to kubeconfig file, that was just created/updated by gcloud get-credentials above
+    String kubeConfigPath = System.getProperty("user.home") + "/.kube/config";
```

Comment:
Do we need to be concerned about malicious users controlling either input for `executeCommand` or `/.kube/config`? The performance cluster shouldn't be world-readable, so I'm not sure how much of an issue this could be.

Reply:
I don't think so. This test runner is designed to be run as a client, not on the same container as our application.